src/coreclr/vm/jitinterface.cpp

```diff
@@ -8898,25 +8898,28 @@ bool CEEInfo::resolveVirtualMethodHelper(CORINFO_DEVIRTUALIZATION_INFO * info)
         // Don't try and devirtualize com interface calls.
         if (pObjMT->IsComObjectType())
         {
-            return nullptr;
+            return false;
         }
 #endif // FEATURE_COMINTEROP
 
         // Interface call devirtualization.
         //
         // We must ensure that pObjMT actually implements the
         // interface corresponding to pBaseMD.
-        if (!pObjMT->CanCastToInterface(pBaseMT))
-        {
-            return false;
-        }
+        bool canCastStraightForward = pObjMT->CanCastToInterface(pBaseMT);
 
         // For generic interface methods we must have context to 
         // safely devirtualize.
+        MethodTable* pOwnerMT = nullptr;
         if (info->context != nullptr)
         {
             TypeHandle OwnerClsHnd = GetTypeFromContext(info->context);
-            MethodTable* pOwnerMT = OwnerClsHnd.GetMethodTable();
+            pOwnerMT = OwnerClsHnd.GetMethodTable();
+
+            if (!canCastStraightForward && !(pOwnerMT->IsInterface() && pObjMT->CanCastToInterface(pOwnerMT)))
```

Comment:
What scenario are you addressing here?

Reply:
Sorry, I found your explanation above. If pOwnerMT isn't an interface, why don't we return false here?

---
In reply to: [570649611](https://github.com/dotnet/runtime/pull/43668#discussion_r570649611) [](ancestors = 570649611)